### PR TITLE
Support simple binary operator for dimension filter

### PR DESCRIPTION
**Pull request by navis**
Currently, dimension filter only supports equal(=) operator. I know it's possible to use other binary operations via javascript filter but it would be way faster and simpler. And also, javascript filter cannot be used with filtered aggregator.

### CODE REVIEW

File: docs/content/querying/filters.md [hunk 1]
```diff
@@ -11,10 +11,12 @@ The simplest filter is a selector filter. The selector filter will match a speci
 The grammar for a SELECTOR filter is as follows:
 
 ``` json
-"filter": { "type": "selector", "dimension": <dimension_string>, "value": <dimension_value_string> }
+"filter": { "type": "selector", "dimension": <dimension_string>, "value": <dimension_value_string>, "operator": <operator> }
 ```
 
-This is the equivalent of `WHERE <dimension_string> = '<dimension_value_string>'`.
+When `operator` is not specified it will be translated to `EQUAL` operator.
```
Comment (b-slim) [inline docs/content/querying/filters.md]: how this is different from what we have like `bound`, `in` and `not`

File: processing/src/main/java/io/druid/segment/filter/SelectorFilter.java [hunk 2]
```diff
@@ -36,26 +40,66 @@
...
   public ImmutableBitmap getBitmapIndex(BitmapIndexSelector selector)
   {
-    return selector.getBitmapIndex(dimension, value);
+    final BitmapFactory factory = selector.getBitmapFactory();
+    final Indexed<String> values = selector.getDimensionValues(dimension);
+    switch (operator) {
+      case GT:
+      case GTE: {
+        int index = values.indexOf(value);
+        int start = index < 0 ? -index - 1 : operator == BinaryOperator.GT ? index + 1: index;
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/segment/filter/SelectorFilter.java]: some javadocs about this logic would make it easier to understand. 
Also, when the value is not present, there are two cases the
1) value is be greater than the current max which means no value match the filter
 2) value is less than the current min, i.e, all values are greater than value

does this needs to be fixed  or am i missing anything ?

Comment (navis) [inline processing/src/main/java/io/druid/segment/filter/SelectorFilter.java]: I'll add testcase for them and check how it's working.

Comment (CHOIJAEHONG1) [inline processing/src/main/java/io/druid/segment/filter/SelectorFilter.java]: Is there any possibility that `selector.getDimensionValues` returns an instance of Indexed<String> which does not honor the semantic of `(minIndex -1)` when the value is absent?

File: processing/src/main/java/io/druid/segment/filter/SelectorFilter.java [hunk 3]
```diff
@@ -66,27 +110,63 @@ public ValueMatcher makeMatcher(ColumnSelectorFactory columnSelectorFactory)
     );
 
     // Missing columns match a null or empty string value and don't match anything else
```
Comment (CHOIJAEHONG1) [inline processing/src/main/java/io/druid/segment/filter/SelectorFilter.java]: I guess this comment should be moved to above the line number 138.

File: processing/src/main/java/io/druid/segment/filter/SelectorFilter.java (hunk 2 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/filter/SelectorFilter.java]: @nishantmonu51 added test cases and confirmed it's ok.

@CHOIJAEHONG1 Yes, it's not documented precisely but BitmapIndexSelector returns index like that. Looks like we need to clarify the semantic of `Indexed.indexof()`, I think.

File: processing/src/main/java/io/druid/segment/filter/SelectorFilter.java (hunk 3 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/filter/SelectorFilter.java]: done

**Comment by b-slim** [comment]
@navis how this is different from what we have like bound, in and not ?

**Comment by navis** [comment]
@b-slim Oh, I haven't noticed bound filter. Well, I think this can be possibly faster than bound filter and can be used with filtered aggregators but it's not support alphanumeric comparator. I cannot insist this should be included then.

**Comment by b-slim** [comment]
@navis sorry,  i don't see how this can enhance the performance, an in general have 1000 ways to do the same thing is not a great thing to have in a project like druid. it is better to keep the code stable and lean as much as possible. Can we make sure that we don't have any duplicate ways of doing thing.

**Comment by navis** [comment]
@b-slim bound filter applies predicate to all values in dictionary but this exploits sorted trait of dictionary and just applies predicate once. I think I can merge two codes into one if you admits to do it.

**Comment by b-slim** [comment]
@navis the performance optimization is independent from the API. The point i want to make, is the fact that we have already told druid user that in order to do bound filter the syntax is `x,y,,,z` therefore we can change that and we are suppose to keep it since it is public. Many users of druid have dependency on our public API. IMHO maintaining the API is more important than optimizing code, especially if users are not complaining about it.

**Comment by b-slim** [comment]
And by the way performance of the bound filters depends directly on the performance of `DimensionPredicateFilter` which is used by many other classes. So any optimization i think it should be done there.

**Comment by navis** [comment]
@b-slim Ok, now it's about API not performance. By word `merge` I meant I'll merge new codes into existing bound filter but it looks like you don't like that. Just one way to do a thing is good but it's not that absolute I think, and we can always deprecate a thing if there is better way(I'm not suggesting this is better) to do it.
And `DimensionPredicateFilter` cannot be optimized as the way I've used because it's limiting the way of filtering to apply the predicate to all values in the dictionary.

**Comment by cheddar** [comment]
If I can weigh in a bit here, performance is always paramount and multiple ways to do the same thing is not necessarily a problem as long as there is a "preferred" way and we are just maintaining one of them for legacy reasons.

On the topic of this change in particular.  I like the general idea of it.  It seems to be trying to simplify the various logic and making it easier to actually do the queries.  I think a lot of it is likely working around the annoyance of how difficult it is to actually register new filter objects.

It can make it a little more difficult if we implement `!=` and other things this way to do optimizations.  Should we expect the "not" filter or should we expect a selector with a `!=`.

If I can redirect the effort a little bit, I think this might be really nice if accompanied by a "new" filter that essentially implements a "WHERE" clause.  There could be a part of the query pipeline added that essentially rewrites the "WHERE" clause into a set of the "primitive" filters that we have today.  If we had that, then we could hide the "selector" filter behind said rewriting.  We could tell users to use that method for their filters and behind the scenes, we can generate whatever ends up being the most performant.

How does that sound?

**Comment by b-slim** [comment]
@navis please Let me explain more what's my point:
1 - This PR is introducing operators like `< `or` >` and doesn't cover all the aspect of it and bound filter does so i don't see how we can have two filters together.
2 -  While it is true from a human perspective it is cool to have one filter that does everything using math like operators, but i think druid API is designed to be automatically generated. So having Filters in the form of POJOs structure is better IMO than having a filter that uses special characters like `!=`. 
 3 - IMO i don't think it is the way to go to have multiple API doing the same thing, if there is no reason to change the API i don't see why we changed.

So know you may ignore 2 and 3, but i don't think 1 can be ignored. If so please explain how this can be ignored or fixed.
Thanks!

**Comment by cheddar** [comment]
I think we can have the best of both worlds here guys.  If we create a filter like

```
{ type: "whereClause",
  "where": "x = billy OR (x = sally and y = 1)" }
```

And then we add a new `QueryRunner` that checks the filter set for instances of the "whereClause" filter and rewrites it into objects like "not", "selector", "bounds", etc. then we can have both worlds.  Downstream logic can be implemented assuming a structured set of objects defining the filter, and we can also support more human-friendly filter statements.

**Comment by b-slim** [comment]
:+1:

**Comment by navis** [comment]
Again, I've already suggested to merging this into existing filter, API or interface, whatever it called at first. And also I like user readable expression(where) which is suggested @cheddar and it's the way to be  done ultimately.

Ok, you win. I'll close this.

**Pull request closed without merging**